cmd/oras/root/manifest/index/update.go

```diff
@@ -0,0 +1,243 @@
+	removeArguments []string
+	tags            []string
+}
+
+func updateCmd() *cobra.Command {
+	var opts updateOptions
+	cmd := &cobra.Command{
+		Use:   "update <name>{:<tag>|@<digest>} {--add/--merge/--remove} {<tag>|<digest>} [...]",
+		Short: "[Experimental] Update and push an image index",
+		Long: `[Experimental] Update and push an image index. All manifests should be in the same repository
+		
+Example - add one manifest and remove two manifests from an index tagged 'latest'. The tag will point to the updated index:
+  oras manifest index update localhost:5000/hello:latest --add linux-amd64 --remove sha256:xxx
+
+Example - update an index by specifying its digest:
+  oras manifest index update localhost:5000/hello@sha256:xxx --add sha256:xxx --remove linux-arm64
+
+Example - remove a manifest and merge manifests from the index tagged as 'index01':
+  oras manifest index update localhost:5000/hello:latest --remove sha256:xxx --merge index01
+
+Example - update an index and tag the updated index:
+  oras manifest index update localhost:5000/hello@sha256:xxx --remove sha256:xxx --tag "latest" --tag "v2.1.0"
+  `,
+		Args: oerrors.CheckArgs(argument.AtLeast(1), "the destination index to update"),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			opts.RawReference = args[0]
+			if err := option.Parse(cmd, &opts); err != nil {
+				return err
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return updateIndex(cmd, opts)
+		},
+	}
+	option.ApplyFlags(&opts, cmd.Flags())
+	cmd.Flags().StringArrayVarP(&opts.addArguments, "add", "", nil, "add manifests to the index")
+	cmd.Flags().StringArrayVarP(&opts.mergeArguments, "merge", "", nil, "merge the manifests of another index")
+	cmd.Flags().StringArrayVarP(&opts.removeArguments, "remove", "", nil, "manifests to remove from the index")
+	cmd.Flags().StringArrayVarP(&opts.tags, "tag", "", nil, "tags for the updated index")
+	return oerrors.Command(cmd, &opts.Target)
+}
+
+func updateIndex(cmd *cobra.Command, opts updateOptions) error {
+	// if no update flag is used, do nothing
+	if !cmd.Flags().Changed("add") && !cmd.Flags().Changed("remove") && !cmd.Flags().Changed("merge") {
+		opts.Println("No update flag is used. There's nothing to update.")
+		return nil
+	}
+	ctx, logger := command.GetLogger(cmd, &opts.Common)
+	target, err := opts.NewTarget(opts.Common, logger)
+	if err != nil {
+		return err
+	}
+	if err := opts.EnsureReferenceNotEmpty(cmd, true); err != nil {
+		return err
+	}
+	index, err := fetchIndex(ctx, target, opts)
+	if err != nil {
+		return err
+	}
+	manifests, err := addManifests(ctx, index.Manifests, target, opts)
+	if err != nil {
+		return err
+	}
+	manifests, err = mergeIndexes(ctx, manifests, target, opts)
+	if err != nil {
+		return err
+	}
+	manifests, err = removeManifests(ctx, manifests, target, opts)
+	if err != nil {
+		return err
+	}
+
+	// media type may be converted to "application/vnd.oci.image.index.v1+json"
```

Comment:
Should we print a warning to users if the original media type was not OCI Image Index?

Reply:
Per https://github.com/oras-project/oras/pull/1476#discussion_r1751299646, it's decided that we don't change the media type to image index.